RavenDB-24649 Measure (approximately) the time elapsed since the last executed query instead of using in-memory values.

maciejaszyk opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-24649

### Additional description

This PR changes the way we measure time from the last query used in maintenance of auto indexes.
Previously, LastQueryTime was reset on index object recreation, which led to never being deleted on databases that are often turned on/off (or e.g. one of the nodes periodically was turned on/off).

This PR introduces new index properties that are measuring and periodically persist elapsed time from last executed query. This allows us to actually delete / set idle an autoindex that has been running and not being queried for a specific threshold. It is also resistant to reset of the db, since it is only measuring time that index is available for querying.

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [x] New feature

### How risky is the change?

- [ ] Low 
- [ ] Moderate 
- [x] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [x] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [ ] No documentation update is needed 

### Testing by Contributor

- [ ] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing
- [x] Existing tests verify the correct behavior

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

maciejaszyk committed:
RavenDB-24649 Measure (approximately) the time elapsed since the last executed query instead of using in-memory values.

maciejaszyk committed:
RavenDB-24649 Studio: added new configuration option.

karmeli87 reviewed (approved):
Handling time might be more tricky than it seems ;)

arekpalinski commented on src/Raven.Server/Documents/Indexes/Index.cs:
```diff
@@ -1913,8 +1924,21 @@ protected void ExecuteIndexing()
+                                    if (waitResult != WaitHandle.WaitTimeout)
+                                        break;
+                                    
+                                    using (_environment.Options.SkipCatastrophicFailureAssertion())
```
Is it really worth it in this case? I'd rather just skip it here, I know we use it above but there is motivation for that then as the comment says,

maciejaszyk replied:
Removed.

arekpalinski commented on src/Raven.Server/Documents/Indexes/Index.cs:
```diff
@@ -174,7 +174,13 @@ public abstract class Index : ITombstoneAware, IDisposable, ILowMemoryHandler
+        /// </summary>
+        private DateTime _lastQueryingTime;
+
+        private TimeSpan _timeElapsedFromLastQuery;
```
Maybe `_elapsedSinceQuery / _elapsedSinceQueried / _elapsedSinceLastQuery` ?

maciejaszyk replied:
Updated

arekpalinski commented on src/Raven.Server/Documents/Indexes/Index.cs:
```diff
@@ -174,7 +174,13 @@ public abstract class Index : ITombstoneAware, IDisposable, ILowMemoryHandler
+        private DateTime _lastQueryingTime;
+
+        private TimeSpan _timeElapsedFromLastQuery;
+        private DateTime _timeElapsedFromLastQueryUpdateDate;
```
Then also update here accordingly

arekpalinski commented on src/Raven.Server/Documents/Indexes/Index.cs:
```diff
@@ -1953,6 +1977,21 @@ protected void ExecuteIndexing()
+        private TimeSpan UpdateElapsedTimeFromLastQuery(DateTime now)
+        {
+            if (now < _timeElapsedFromLastQueryUpdateDate)
+                return _timeElapsedFromLastQuery;
```
Why do we need this?

maciejaszyk replied:
There is a possible race between `Update` and `MarkQueried` methods. If the update loses the race, we have nothing to do since the actual values are from the "future" from the perspective of the caller.

arekpalinski commented on src/Raven.Server/Documents/Indexes/Index.cs:
```diff
@@ -1953,6 +1977,21 @@ protected void ExecuteIndexing()
+            var timeElapsedFromLastQuery = _timeElapsedFromLastQuery;
+            timeElapsedFromLastQuery += elapsed;
+
+
```
Doubled empty line

maciejaszyk replied:
removed

arekpalinski commented on src/Raven.Server/Documents/Indexes/Index.cs:
```diff
@@ -1953,6 +1977,21 @@ protected void ExecuteIndexing()
             }
         }
 
+        private TimeSpan UpdateElapsedTimeFromLastQuery(DateTime now)
```
Please consider that it can be called concurrently with `MarkQueried`. This can potentially lead to skipping / overwriting some real elapsed querying time.

maciejaszyk replied:
Handled.

arekpalinski commented on src/Raven.Server/Documents/Indexes/IndexStorage.cs:
```diff
@@ -1255,6 +1286,8 @@ internal sealed class IndexSchema
 
             public static readonly Slice LastIndexingTimeSlice;
+            
+            public static readonly Slice ElapsedTimeFromLastQuerySlice;
```
Please rename accordingly for consistency. Please see one of above comment

maciejaszyk replied:
Renamed and boxed.

arekpalinski commented on src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs:
```diff
@@ -561,44 +553,36 @@ private static string GetCommandId(Dictionary<string, long> dic)
-                    if (difference == TimeSpan.Zero && lowestDatabaseUpTime > kvp.Value)
-                        difference = kvp.Value;
-                }
+                var smallestTimeElapsedInCluster = kvp.Value;
```
How do we ensure that it's really the smallest. It's not obvious so please add a comment.

arekpalinski commented on src/Raven.Server/Config/Categories/IndexingConfiguration.cs:
```diff
@@ -597,6 +597,13 @@ public virtual bool RunInMemory
+        [DefaultValue(10)]
+        [TimeUnit(TimeUnit.Minutes)]
+        [IndexUpdateType(IndexUpdateType.None)]
+        [ConfigurationEntry("Indexing.IntervalOfStoringElapsedTimeFromLastQueryInMin", ConfigurationEntryScope.ServerWideOrPerDatabaseOrPerIndex)]
```
Typically we use `...IntervalInMin` suffix in configuration names so maybe `Indexing.ElapsedSinceLastQueryPersistIntervalInMin`

maciejaszyk replied:
done

arekpalinski reviewed (commented):
The tests are missing. I can think about the following:

- checking that after the restart we have a correct value of elapsed since last query
- backup / restore - after the restore (snapshot) it won't attempt to delete auto index immediately

maciejaszyk committed:
RavenDB-24649

- Thread safe time tracker
- Encapsulated all last query time / elapsed in separate file
- Changed naming conventions
- Added tests

arekpalinski commented:
Please check the test failure:
https://jenkins.hibernatingrhinos.com/job/PR_Tests/16039/testReport/junit/SlowTests.Issues/RavenDB_24649/ElapsedSinceQueriedTimeIsPersistedAndIncreased/

maciejaszyk committed:
RavenDB-24649 Fix flaky test

arekpalinski commented:
Please check: https://jenkins.hibernatingrhinos.com/job/PR_Tests/16069/testReport/SlowTests.Issues/RavenDB_24649/ElapsedSinceQueriedTimeIsPersistedAndIncreased/
